packages/core/src/codecatalyst/devEnv.ts

```diff
@@ -11,6 +11,45 @@ import { Timeout } from '../shared/utilities/timeoutUtils'
 import { showMessageWithCancel } from '../shared/utilities/messages'
 import { isCloud9 } from '../shared/extensionUtilities'
 import { getLogger } from '../shared/logger'
+import { CodeCatalystAuthenticationProvider } from './auth'
+import { getThisDevEnv } from './model'
+
+let didStartDevEnvActivityHeartbeat: boolean = false
+/**
+ * Trys to start the Dev Env Activity Heartbeat to keep the Dev Env from timing out.
+ * If this successfully starts, subsequent calls to this function will do nothing.
+ *
+ * ---
+ * Why do we need this?
+ *
+ * We need this function since there are times the heartbeat mechanism cannot be started. Eg auth
+ * is not connected in a Dev Env. So we need to try again elsewhere.
+ */
+export async function tryStartDevEnvActivityHeartbeat(authProvider: CodeCatalystAuthenticationProvider) {
+    if (didStartDevEnvActivityHeartbeat) {
+        return
+    }
+    const thisDevenv = (await getThisDevEnv(authProvider))?.unwrapOrElse(err => {
+        getLogger().warn('codecatalyst: failed to get current Dev Enviroment: %s', err)
+        return undefined
+    })
+
+    if (!thisDevenv) {
+        getLogger().verbose('codecatalyst: no dev env to start heartbeat')
+        return
+    }
+
+    const maxInactivityMinutes = thisDevenv.summary.inactivityTimeoutMinutes
+    const devEnvClient = thisDevenv.devenvClient
+    const devEnvActivity = await DevEnvActivity.instanceIfActivityTrackingEnabled(devEnvClient)
+    if (shouldTrackUserActivity(maxInactivityMinutes) && devEnvActivity) {
```

Comment:
In theory, any `await` call may park the current execution and re-enter `tryStartDevEnvActivityHeartbeat` from e.g. `onDidChangeActiveConnection` during startup.  Even if that's unlikely or currently impossible, it's always a good idea to make the expectations as clear as possible. Thus we should check `didStartDevEnvActivityHeartbeat` just before we create the resource(s).


```suggestion
    if (shouldTrackUserActivity(maxInactivityMinutes) && devEnvActivity && !didStartDevEnvActivityHeartbeat) {
```

Reply:
Good find, I noticed this was getting called before fully completing. I've refactored a bit, but updated to use the `shared()` helper function to essentially debounce any subsequent calls that can occur.